### PR TITLE
Adding health check for streams

### LINKED ISSUE #120: Add KafkaStreams Health Class
Putting this here for backlog for me (anyone) to add

```java

import io.micronaut.configuration.kafka.config.AbstractKafkaConfiguration;
import io.micronaut.context.annotation.Requires;
import io.micronaut.health.HealthStatus;
import io.micronaut.management.health.aggregator.HealthAggregator;
import io.micronaut.management.health.indicator.HealthIndicator;
import io.micronaut.management.health.indicator.HealthResult;
import io.reactivex.Flowable;
import org.apache.kafka.streams.KafkaStreams;
import org.reactivestreams.Publisher;

import javax.inject.Singleton;
import java.util.Collection;

@Singleton
@Requires(classes = HealthIndicator.class)
@Requires(property = AbstractKafkaConfiguration.PREFIX + ".health.enabled", value = "true", defaultValue = "true")
public class KafkaStreamsHealth implements HealthIndicator {

    private final Collection<KafkaStreams> streams;
    private final HealthAggregator<?> healthAggregator;

    private static final String NAME = "kafkaStreams";

    public KafkaStreamsHealth(final Collection<KafkaStreams> streams, final HealthAggregator<?> healthAggregator) {
        this.streams = streams;
        this.healthAggregator = healthAggregator;
    }

    @Override
    public Publisher<HealthResult> getResult() {
        return Flowable.fromIterable(streams)
                .map(kafkaStreams -> {
                    try {
                        if (kafkaStreams.state().isRunning()) {
                            return HealthResult.builder(kafkaStreams.toString(), HealthStatus.UP).build();
                        } else {
                            return HealthResult.builder(kafkaStreams.toString(), HealthStatus.DOWN).build();
                        }
                    } catch (Exception e) {
                        return HealthResult.builder("Unknown", HealthStatus.DOWN).build();
                    }
                })
                .flatMap(healthResult -> healthAggregator.aggregate(NAME, Flowable.just(healthResult)));

    }
}

```

### LINKED ISSUE #132: Micronaut tests use embedded kafka AND test container
It looks like the way the tests are setup there is a kafka container from the test containers project started and not used when in fact the embedded kafka is started and used by Micronaut.

### PR DESCRIPTION
**Pull request by ctoestreich**
* fixes #120 
* fixes #132 for streams module

I had to downgrade the test container version due to issue with docker java container https://github.com/docker-java/docker-java/pull/1392

### CODE REVIEW

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/KafkaStreamsFactory.java [hunk 1]
```diff
@@ -42,11 +41,13 @@
 @Factory
 public class KafkaStreamsFactory implements Closeable {
 
-    private final Collection<KafkaStreams> streams = new ConcurrentLinkedDeque<>();
+    private final Map<KafkaStreams, ConfiguredStreamBuilder> streams = new ConcurrentHashMap<>();
```
Comment (brianwyka) [inline kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/KafkaStreamsFactory.java:44]: Does `KafkaStreams` implement `hashCode()`

Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/KafkaStreamsFactory.java:44]: Why do you need access to the builder in this map.  Aren't all the configurations that the `KafkaStreams` was built with available there?

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+     * for each configured stream as the primary health check name.
+     *
+     * @return Health Result Aggregate
+     */
+    @Override
+    public Publisher<HealthResult> getResult() {
+        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(kafkaStreamsFactory.getStreams().keySet())
+                .map(kafkaStreams -> {
+                    String applicationId = getApplicationId(kafkaStreams);
+                    try {
```
Comment (brianwyka) [inline kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:81]: Perhaps use `filter`, `defaultIfEmpty` and `onErrorReturn` instead?

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    /**
+     * Get default name if application.id and client.id are not present.
+     *
+     * @param kafkaStreams the kafka stream
+     * @return The name
+     */
+    private String getDefaultStreamName(final KafkaStreams kafkaStreams) {
+        if (kafkaStreams != null) {
+            return Optional.ofNullable(kafkaStreams.localThreadsMetadata())
+                    .map(threadMetadataSet -> threadMetadataSet.stream().findFirst())
```
Comment (brianwyka) [inline kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:171]: method reference can be used here

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    /**
+     * Get the health result of the streams.  Will attempt to interrogate
+     * details of each stream as well.  The application.id will be used
+     * for each configured stream as the primary health check name.
+     *
+     * @return Health Result Aggregate
+     */
+    @Override
+    public Publisher<HealthResult> getResult() {
+        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(kafkaStreamsFactory.getStreams().keySet())
```
Comment (brianwyka) [inline kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:78]: Do we want one health aggregated health result or many?  if we do `fromIterable`, we will get many.  You can flatten them into one result if that is what actually is desired.

Comment (ctoestreich) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:78]: We want a high level single status that comprises of on status per kafkastream once the aggregator with name at the end

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 2 above)
Comment (ctoestreich) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:81]: yeah I am not the rx guru like you. `maybe(that would be better?).defaultIfEmpty("@bwyka does code change").onErrorReturn("do not change the code at all?")`

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 3 above)
Comment (ctoestreich) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:171]: Best I can do is this

```
return Optional.ofNullable(kafkaStreams.localThreadsMetadata())
                    .map(Collection::stream)
                    .flatMap(Stream::findFirst)
                    .map(ThreadMetadata::threadName)
                    .orElse(kafkaStreams.toString());
```

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 4 above)
Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:78]: If that's the case, then you will want to do `Flowable.just` with the iterable as the parameter, and then you can have a list of generated details pertaining to the individual `HealthResult`.  I did something similar for hystrix as seen here:

```java
@Singleton
--
  | @AllArgsConstructor
  | @Requires(beans = HystrixInterceptor.class)
  | public class HystrixHealthIndicator implements HealthIndicator {
  |  
  | private static final String HEALTH_RESULT_NAME = "hystrixCircuitBreaker";
  | private static final String CIRCUIT_OPEN = "CIRCUIT_OPEN";
  | private static final String CIRCUIT_CLOSED = "CIRCUIT_CLOSED";
  |  
  | /** {@inheritDoc} */
  | @Override
  | public Publisher<HealthResult> getResult() {
  | return Flowable.fromIterable(HystrixCommandMetrics.getInstances())
  | .flatMapMaybe(HystrixHealthIndicator::aggregateCircuitBreakers)
  | .defaultIfEmpty(HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UP))
  | .map(HealthResult.Builder::build);
  | }
  |  
  | /**
  | * Aggregate circuit breakers into health results
  | *
  | * @param hystrixCommandMetrics the hystrix command metrics
  | * @return the aggregated circuit breaker health results
  | */
  | private static Maybe<HealthResult.Builder> aggregateCircuitBreakers(final HystrixCommandMetrics hystrixCommandMetrics) {
  | return Single.fromCallable(hystrixCommandMetrics::getCommandKey)
  | .map(HystrixCircuitBreaker.Factory::getInstance)
  | .map(Optional::ofNullable)
  | .filter(Optional::isPresent)
  | .map(Optional::get)
  | .map(hystrixCircuitBreaker -> HealthResult.builder(String.format("%s-%s", HEALTH_RESULT_NAME, hystrixCommandMetrics.getCommandKey().name()))
  | .status(deriveCircuitBreakerHealthStatus(hystrixCommandMetrics, hystrixCircuitBreaker))
  | .details(hystrixCommandMetrics.getHealthCounts()))
  | .defaultIfEmpty(HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UP))
  | .onErrorReturn(e -> HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UNKNOWN).exception(e));
  | }
  |  
  | /**
  | * Derive the {@link HealthStatus} for the circuit based on whether or not its open
  | *
  | * @param hystrixCommandMetrics the hystrix command metrics
  | * @param hystrixCircuitBreaker the hystrix circuit breaker
  | * @return the health status
  | */
  | private static HealthStatus deriveCircuitBreakerHealthStatus(final HystrixCommandMetrics hystrixCommandMetrics, final HystrixCircuitBreaker hystrixCircuitBreaker) {
  | if (hystrixCircuitBreaker.isOpen()) {
  | return new HealthStatus(CIRCUIT_OPEN, null, false, hystrixCommandMetrics.getHealthCounts().getErrorPercentage() * 100);
  | }
  | return new HealthStatus(CIRCUIT_CLOSED, null, true, null);
  | }
  |  
  | }


```

Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:78]: Yeah, I think you are all set here then.  Overlooked the `aggregate` at the bottom.

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 2 above)
Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:81]: At your command:

```java
    @Override
    public Publisher<HealthResult> getResult() {
        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(Collections.singleton(new KafkaStreams((Topology) null, (Properties) null)))
                .map(kafkaStreams -> Pair.of(getApplicationId(kafkaStreams), kafkaStreams))
                .flatMap(pair -> Flowable.just(pair)
                        .filter(p -> p.getValue().state().isRunning())
                        .map(p -> HealthResult.builder(p.getKey(), HealthStatus.UP)
                                .details(buildDetails(p.getValue())))
                        .defaultIfEmpty(HealthResult.builder(pair.getKey(), HealthStatus.DOWN)
                                .details(Collections.singletonMap("error", "The processor is down")))
                        .onErrorReturn(e -> HealthResult.builder(pair.getKey(), HealthStatus.DOWN)
                                .details(buildDownDetails(e, pair.getValue().state()))))
                .map(HealthResult.Builder::build);
        return healthAggregator.aggregate(NAME, kafkaStreamHealth);
    }
```

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 3 above)
Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:171]: 👍

Comment (brianwyka) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:171]: Or even:

```java
    private String getDefaultStreamName(final KafkaStreams kafkaStreams) {
        return Optional.ofNullable(kafkaStreams)
                .map(KafkaStreams::localThreadsMetadata)
                .map(Collection::stream)
                .flatMap(Stream::findFirst)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(ThreadMetadata::threadName)
                .orElse("unidentified");
    }
```

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/KafkaStreamsFactory.java (hunk 1 above)
Comment (ctoestreich) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/KafkaStreamsFactory.java:44]: Unfortunately the builder has the properties accessible that were used to build.  The configuration properties are NOT accessible from the KafkaStreams object directly as it is `private final StreamsConfig config;private final StreamsConfig config;`. The ONLY way to pluck out the application.id or client.id is to tie th streams to the configuration builder.  This is what Spring Boot does as well.  I used their pattern as a guide.

File: kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java (hunk 4 above)
Comment (ctoestreich) [reply kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java:78]: 👍

**Comment by graemerocher** [comment]
Thanks for the contribution

**Pull request merged**